execution-plane/arcon/arcon_state/src/rocks/vec_ops.rs

```diff
@@ -0,0 +1,229 @@
+    ) -> Result<BoxedIteratorOfResult<'_, T>> {
+        let key = handle.serialize_metakeys()?;
+        if let Some(serialized) = self.get(handle.id, &key)? {
+            // TODO: this would be nicer with generators :(
+            let mut reader = &serialized[..];
+            let origin = reader.as_ptr() as usize;
+            // we ignore the length, but we do the deserialization anyway to advance the reader
+            let _: usize = fixed_bytes::deserialize_from(&mut reader)?;
+            // this is safe because we're counting bytes
+            let mut consumed = reader.as_ptr() as usize - origin;
+
+            let iter = iter::from_fn(move || {
+                // We cannot use the borrow from the outside, because it points to the local fn
+                // variable. We have to force `serialized` to move into the closure. We have to
+                // manually keep track of the number of consumed bytes, because otherwise we'd
+                // have a self-referential struct.
+                let mut reader = &serialized[consumed..];
+
+                if !reader.is_empty() {
+                    let res = protobuf::deserialize_from(&mut reader);
+                    consumed = reader.as_ptr() as usize - origin;
+                    Some(res)
+                } else {
+                    None
+                }
+            });
+
+            Ok(Box::new(iter))
+        } else {
+            Ok(Box::new(iter::empty()))
+        }
+    }
+
+    fn vec_set<T: Value, IK: Metakey, N: Metakey>(
+        &mut self,
+        handle: &Handle<VecState<T>, IK, N>,
+        value: Vec<T>,
+    ) -> Result<()> {
+        let key = handle.serialize_metakeys()?;
+        let raw_serialized_len: usize = value
+            .iter()
+            .flat_map(|x| protobuf::size_hint(x).into_iter())
+            .sum();
+        let cap = <usize as FixedBytes>::SIZE + raw_serialized_len;
+
+        let mut storage = Vec::with_capacity(cap);
+        fixed_bytes::serialize_into(&mut storage, &value.len())?;
+        for elem in value {
+            protobuf::serialize_into(&mut storage, &elem)?;
+        }
+
+        self.put(handle.id, key, storage)
+    }
+
+    fn vec_add_all<T: Value, IK: Metakey, N: Metakey>(
+        &mut self,
+        handle: &Handle<VecState<T>, IK, N>,
+        values: impl IntoIterator<Item = T>,
+    ) -> Result<()> {
+        let backend = self.initialized_mut()?;
+        let key = handle.serialize_metakeys()?;
+
+        // figuring out the correct capacity would require iterating through `values`, but we
+        // cannot really consume the `values` iterator twice, so just preallocate a bunch of bytes
+        let mut serialized = Vec::with_capacity(256);
+
+        // reserve space for the length
+        fixed_bytes::serialize_into(&mut serialized, &0usize)?;
+        let mut len = 0usize;
+
+        for elem in values {
+            len += 1;
+            protobuf::serialize_into(&mut serialized, &elem)?;
```

Comment:
What will happen if the amount of bytes go above 256 here in this loop? does the ``protobuf::serialize_into```allocate more memory into the memory block?

Reply:
Yes, `BufMut` impl for `Vec<u8>` just grows the vec if the capacity is reached.